### ISSUE TITLE
Jupyterlab, Rstudio and VSCode do not run as non root in Kubeflow 1.8

### ISSUE THREAD

**Issue by juliusvonkohout** [kubeflow/kubeflow#5808]
/kind bug

@DavidSpek 

```
[julius@fedora 1.3]$ kubectl -n kubeflow-user logs pod/julius1-0
s6-overlay-preinit: fatal: unable to mkdir /var/run/s6: Permission denied
```

It also does not work with podman

```
podman run --user 1:0 public.ecr.aws/j1r0q0g6/notebooks/notebook-servers/jupyter-scipy:v1.3.0-rc.0 

[s6-init] making user provided files available at /var/run/s6/etc...exited 0.
[s6-init] ensuring user provided files have correct perms...s6-chown: fatal: unable to chown /var/run/s6/etc/cont-init.d/01-copy-tmp-home: Operation not permitted
s6-chmod: fatal: unable to change mode of /var/run/s6/etc/cont-init.d/01-copy-tmp-home: Operation not permitted
s6-chown: fatal: unable to chown /var/run/s6/etc/services.d/jupyterlab/run: Operation not permitted
s6-chmod: fatal: unable to change mode of /var/run/s6/etc/services.d/jupyterlab/run: Operation not permitted
exited 0.
[fix-attrs.d] applying ownership & permissions fixes...
[fix-attrs.d] done.
[cont-init.d] executing container initialization scripts...
[cont-init.d] 01-copy-tmp-home: executing... 
cp: cannot access '/tmp_home/jovyan/.jupyter': Permission denied
[cont-init.d] 01-copy-tmp-home: exited 1.
[cont-init.d] done.
[services.d] starting services
s6-supervise (child): fatal: unable to exec run: Permission denied
s6-supervise jupyterlab: warning: unable to spawn ./run - waiting 10 seconds
[services.d] done.
s6-supervise (child): fatal: unable to exec run: Permission denied
s6-supervise jupyterlab: warning: unable to spawn ./run - waiting 10 seconds
s6-supervise (child): fatal: unable to exec run: Permission denied
s6-supervise jupyterlab: warning: unable to spawn ./run - waiting 10 seconds
s6-supervise (child): fatal: unable to exec run: Permission denied
```

The user should be set in the statefulset itself if the container is not good enough to run as any user

```
podman run --user 1000:0 public.ecr.aws/j1r0q0g6/notebooks/notebook-servers/jupyter-scipy:v1.3.0-rc.0 
[s6-init] making user provided files available at /var/run/s6/etc...exited 0.
[s6-init] ensuring user provided files have correct perms...s6-chown: fatal: unable to chown /var/run/s6/etc/cont-init.d/01-copy-tmp-home: Operation not permitted
s6-chown: fatal: unable to chown /var/run/s6/etc/services.d/jupyterlab/run: Operation not permitted
exited 0.
[fix-attrs.d] applying ownership & permissions fixes...
[fix-attrs.d] done.
[cont-init.d] executing container initialization scripts...
[cont-init.d] 01-copy-tmp-home: executing... 
[cont-init.d] 01-copy-tmp-home: exited 0.
[cont-init.d] done.
[services.d] starting services
[services.d] done.
[I 2021-04-06 11:16:26.827 ServerApp] jupyterlab | extension was successfully linked.
[I 2021-04-06 11:16:26.834 ServerApp] Writing notebook server cookie secret to /home/jovyan/.local/share/jupyter/runtime/jupyter_cookie_secret
[I 2021-04-06 11:16:27.103 ServerApp] nbclassic | extension was successfully linked.
[W 2021-04-06 11:16:27.186 ServerApp] All authentication is disabled.  Anyone who can connect to this server will be able to run code.
[I 2021-04-06 11:16:27.203 LabApp] JupyterLab extension loaded from /opt/conda/lib/python3.8/site-packages/jupyterlab
[I 2021-04-06 11:16:27.203 LabApp] JupyterLab application directory is /opt/conda/share/jupyter/lab
[I 2021-04-06 11:16:27.207 ServerApp] jupyterlab | extension was successfully loaded.
[I 2021-04-06 11:16:27.224 ServerApp] nbclassic | extension was successfully loaded.
[I 2021-04-06 11:16:27.225 ServerApp] Serving notebooks from local directory: /home/jovyan
[I 2021-04-06 11:16:27.225 ServerApp] Jupyter Server 1.4.1 is running at:
[I 2021-04-06 11:16:27.225 ServerApp] http://1ecd45fc4e8d:8888/lab
[I 2021-04-06 11:16:27.225 ServerApp]  or http://127.0.0.1:8888/lab
[I 2021-04-06 11:16:27.225 ServerApp] Use Control-C to stop this server and shut down all kernels (twice to skip confirmation).

```

**Comment by DavidSpek** [comment]
@juliusvonkohout Just for clarity, I've made sure that nothing inside the container runs as root. For this reason, the following part of the error message is normal and not actually a problem:
```
[s6-init] making user provided files available at /var/run/s6/etc...exited 0.
[s6-init] ensuring user provided files have correct perms...s6-chown: fatal: unable to chown /var/run/s6/etc/cont-init.d/01-copy-tmp-home: Operation not permitted
s6-chmod: fatal: unable to change mode of /var/run/s6/etc/cont-init.d/01-copy-tmp-home: Operation not permitted
s6-chown: fatal: unable to chown /var/run/s6/etc/services.d/jupyterlab/run: Operation not permitted
s6-chmod: fatal: unable to change mode of /var/run/s6/etc/services.d/jupyterlab/run: Operation not permitted
exited 0.
[fix-attrs.d] applying ownership & permissions fixes...
[fix-attrs.d] done.
```

The `/tmp_home/jovyan/` should be owned by UID 1000, GID 100, so I am not sure why you are getting permission denied for the copying. Looking at the rest of the error outputs, I've only seen this when the init and service scripts in `/etc/cont-init.d/` and `/etc/services.d` are not owned by `jovyan:users` (UID 1000, GID 100). I'm not sure why this is happening with podman as I've never used it. 

Can you try setting the environment variable `S6_READ_ONLY_ROOT=1` and seeing if that resolves any of the problems?

**Comment by juliusvonkohout** [comment]
Ok,then UID 1000, GID 100 must be set in the securitycontext of the pod. Otherwise the common default is used, which is usually not 1000:100

```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: test1
  namespace: kubeflow-user
spec:
  replicas: 1
  selector:
    matchLabels:
      app: test1
  template:
    metadata:
      labels:
        app: test1
    spec:
      containers:
      - env:
        - name: S6_READ_ONLY_ROOT
          value: "1"
        image: public.ecr.aws/j1r0q0g6/notebooks/notebook-servers/jupyter-scipy:v1.3.0-rc.0
        imagePullPolicy: IfNotPresent
        name: jupyter-scipy
        securityContext:
          runAsGroup: 100
          runAsUser: 1000
```

```
[julius@fedora 1.3]$ kubectl -n kubeflow-user logs pod/test1-65d8c66b4c-dx4q9
s6-overlay-preinit: fatal: unable to mkdir /var/run/s6: Permission denied
```

I am using these security settings

```
apiVersion: policy/v1beta1
kind: PodSecurityPolicy
metadata:
  name: kubeflow-restricted-psp
spec:
  privileged: false
  allowPrivilegeEscalation: false
  hostPID: false
  hostIPC: false
  hostNetwork: false
  # WARNING HOSTPORTS
  hostPorts:
  - min: 1024
    max: 65535
  requiredDropCapabilities:
  - ALL
  fsGroup:
    ranges:
    - max: 65535
      min: 1
    rule: MustRunAs
  runAsGroup:
    ranges:
    - max: 65535
      min: 0
    rule: MustRunAs
  runAsUser:
    ranges:
    - max: 65535
      min: 1000
    rule: MustRunAs
    #rule: MustRunAsNonRoot
  seLinux:
    rule: RunAsAny
  supplementalGroups:
    ranges:
    - max: 65535
      min: 1
    rule: MustRunAs
  volumes:
  - configMap
  - emptyDir
  - projected
  - secret
  - downwardAPI
  - persistentVolumeClaim
 ```

**Comment by DavidSpek** [comment]
@julienstroheker I currently don't have the resources to look into this much, can you confirm the issue is also still present with podman? Then I might be able to look into this locally. I would like to note that these images are provided as examples at this point. While the README for these images has not yet been merged, you can take a look here to get more information: https://github.com/kubeflow/kubeflow/pull/5694/files#diff-653fc34c1739e3016f7862a72acb34b43170bbaed9177528747b81878fa0e5c1

**Comment by juliusvonkohout** [comment]
The s6 init system is broken by design, but there might be a workaround

```
[julius@fedora 1.3]$ podman run -p 8888:8888 --user 1000:100 --entrypoint='/bin/bash' public.ecr.aws/j1r0q0g6/notebooks/notebook-servers/jupyter-scipy:v1.3.0-rc.0 -c 'ls -lah /var/run'
lrwxrwxrwx. 1 root root 4 Feb 17 01:04 /var/run -> /run
```

This is also documented here https://github.com/just-containers/s6-overlay#read-only-root-filesystem

```
... first copy user-provided files to its work area in /var/run/s6 before attempting to change permissions.
This of course assumes that at least /var is backed by a writeable filesystem with execute privileges. This could be done with a tmpfs filesystem as follows:
```
So one could try the following in the dockerfile

```
mkdir -p /var/run/s6
chown -R ${NB_USER}:users /var/run/s6
```

I will try it.

**Comment by DavidSpek** [comment]
`/var/run/s6` is created on container startup as far as I know, so the folder is owned by the creator. Otherwise, it would also fail to start when using docker or containerd as s6 is not executed as root.

**Comment by DavidSpek** [comment]
As I suspected, the `/var/run/s6` directory does not exist in the container image as it is created on startup so it can't be `chown`ed in the Dockerfile, and it can't be done at boot as S6 isn't run as root. 
Could you try creating that directory in the Dockerfile and running `chown` to see if that allows it to start properly?

**Comment by juliusvonkohout** [comment]
> `/var/run/s6` is created on container startup as far as I know, so the folder is owned by the creator. Otherwise, it would also fail to start when using docker or containerd as s6 is not executed as root.

Yes, and a non-root user should not be able to write to it.

```
FROM public.ecr.aws/j1r0q0g6/notebooks/notebook-servers/jupyter-scipy:v1.3.0-rc.0

USER root
RUN mkdir -p /var/run/s6 && chown -R ${NB_USER}:users /var/run/s6
USER $NB_UID
```
Is not working because of that

```
[julius@fedora 1.3]$ podman run -p 8888:8888 --user 1000:100 --entrypoint='/bin/bash' public.ecr.aws/j1r0q0g6/notebooks/notebook-servers/jupyter-scipy:v1.3.0-rc.0 -c 'mkdir /var/run/s6'
mkdir: cannot create directory ‘/var/run/s6’: Permission denied
```
They are using root rights via setuid in S6 AFAIK to create the directory at startup. Obviously this privilege escalation is forbidden in a secure cluster.

I will try to mount an emptydir below it.

**Comment by juliusvonkohout** [comment]
This works and shows how broken s6 is security-wise. 

```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: test1
  namespace: kubeflow-user
spec:
  replicas: 1
  selector:
    matchLabels:
      app: test1
  template:
    metadata:
      labels:
        app: test1
    spec:
      containers:
      - env:
        - name: S6_READ_ONLY_ROOT
          value: "1"
        image: public.ecr.aws/j1r0q0g6/notebooks/notebook-servers/jupyter-scipy:v1.3.0-rc.0
        imagePullPolicy: IfNotPresent
        name: jupyter-scipy
        securityContext:
          runAsGroup: 100
          runAsUser: 1000
        volumeMounts:
        - mountPath: /var/run/s6
          name: s6
      volumes:
      - name: s6
        emptyDir: {}
```

Maybe it is possible to change the default s6 working directory from /var/run/s6 to /tmp//var/run/s6

**Comment by juliusvonkohout** [comment]
So a workaround would be to add

1. securitycontext 1000:100
2. emptydir volume
3. volume mount to /var/run/s6

as shown above to the statefulset.

**Comment by juliusvonkohout** [comment]
> I don't think supervisord can properly handle multiple processes while running in a docker container or make it easy for people to run startup scripts, but that is beside the point here. Is the `S6_READ_ONLY_ROOT` variable still needed when you mount the emptydir volume to `/run/s6` (`/var/run` is a symlink to `/run`)? Also, there fixes aren't actually for the notebook images themselves, so I'm not sure where you want to create a PR for this.
> 
> Also, this mounting of an emptydir to either `/var/run` or `/run` might break RStudio.

I am mounting the emptydir under /var/run/s6 i do not see how this should affect /var/run/rstudio-server .
Yes the changes must be done in the statefulset that runs the pods. There you can set the securitycontext etc. I have to check where that is stored in the kubeflow repository. If you have a hint please tell me.

**Comment by andreas-schliebitz** [comment]
Hi, I'am also experiencing the exact same issue while trying to deploy the `jupyter-scipy:v1.3.0-rc.1` Notebook Server via the KubeFlow 1.3 WebUI into the default `kubeflow-user-example-com` namespace on an OpenShift 4.7 cluster:

```
[s6-init] making user provided files available at /var/run/s6/etc...exited 0.
[s6-init] ensuring user provided files have correct perms...s6-chown: fatal: unable to chown /var/run/s6/etc/cont-init.d/01-copy-tmp-home: Operation not permitted
s6-chmod: fatal: unable to change mode of /var/run/s6/etc/cont-init.d/01-copy-tmp-home: Operation not permitted
s6-chown: fatal: unable to chown /var/run/s6/etc/services.d/jupyterlab/run: Operation not permitted
s6-chmod: fatal: unable to change mode of /var/run/s6/etc/services.d/jupyterlab/run: Operation not permitted
exited 0.
[fix-attrs.d] applying ownership & permissions fixes...
[fix-attrs.d] done.
[cont-init.d] executing container initialization scripts...
[cont-init.d] 01-copy-tmp-home: executing...
cp: cannot access '/tmp_home/jovyan/.jupyter': Permission denied
[cont-init.d] 01-copy-tmp-home: exited 1.
[cont-init.d] done.
[services.d] starting services
s6-supervise (child): fatal: unable to exec run: Permission denied
s6-supervise jupyterlab: warning: unable to spawn ./run - waiting 10 seconds
[services.d] done.
s6-supervise (child): fatal: unable to exec run: Permission denied
s6-supervise jupyterlab: warning: unable to spawn ./run - waiting 10 seconds
```

After a while the Notebook status changes to `Running` but trying to `CONNECT` results in a 503 `upstream connect error or disconnect/reset before headers. reset reason: connection failure`.

**Comment by juliusvonkohout** [comment]
@andreas-schliebitz you might want to check out https://github.com/kubeflow/manifests/pull/1875 if you are on Openshift.

In the end I just wrote my own jupyterlab with code-server and RStudio all included in one image.

**Comment by DavidSpek** [comment]
It wouldn't be difficult to extend the notebook controller to add the security context and to mount an empty volume to `/var/run/s6`. I think the main question is if this should be set for all kubernetes cluster, or if it this should be done when detecting it is running on OpenShift.

**Comment by juliusvonkohout** [comment]
Definitely for all,because it is not Openshift specific. On a plain Kubernetes with pod security policies you have exactly the same problem.

**Comment by DavidSpek** [comment]
@juliusvonkohout Is this in terms of adding the security context? Do you know if it is possible to run the notebook image without the security context added but with the empty volume mounted to /var/run/s6? I was planning on creating an experimental version of the notebook controller today for people running into these issues, but I’m not sure about the consequences of the security context and file permissions. 

Do you know if a PVC used for a notebook used with this security context will work properly within a pipeline?

**Comment by juliusvonkohout** [comment]
> So a workaround would be to add
> 
> 1. securitycontext 1000:100
> 2. emptydir volume
> 3. volume mount to /var/run/s6
> 4. S6_READ_ONLY_ROOT env
> 
> as shown above to the statefulset. ShouId i create a pull request? I use supervisord myself that does not have such problems.

As far as I remember all of the four points above were needed, but I have to check again. I really hope that fsgroup 100 is not needed, because that is a problem on openshift. So I need to investigate a bit more.

**Comment by DavidSpek** [comment]
Right, it was my understanding that Kubernetes would run the container with the permissions used within the container. Forcing a security context, and therefore file permissions can break things very easily in different environments. If you could check if a notebook image can run without the security context that would be great.

Regarding the fsgroup, I believe the notebook-controller is already setting that in some places. I'm also not sure how the fsgroup will relate to the user and group set in the security context, so I'll need to look into this. Is there a particular reason why fsgroup 100 is a problem on OpenShift?

**Comment by andreas-schliebitz** [comment]
Interestingly, the default TensorFlow image ([quay.io/kubeflow/tf-notebook-image:v0.7.0](quay.io/kubeflow/tf-notebook-image:v0.7.0)) used in an older KubeFlow 1.2 installation works in 1.3. At least there is no 503 error occurring after hitting `CONNECT` but the Jupyter UI is displayed as expected. For some unknown reason I then get a WebSocket disconnect to the Kernel when trying to launch a cell. However, this might be an issue on my end and not related to KubeFlow, will investigate.

**Comment by DavidSpek** [comment]
@andreas-schliebitz That issue is most likely browser related. I've noticed that this can usually be resolved by using incognito mode, restarting the browser entirely or trying to use a different browser.
The fact that this older notebook image does work makes me think that all that is necessary to make the new notebook images work on OpenShift is the addition of the empty volume mounted at `/var/run/s6` and the environment variable `6_READ_ONLY_ROOT=1`.

Would you be able to test if the following deployment works correctly?

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: test1
  namespace: kubeflow-user
spec:
  replicas: 1
  selector:
    matchLabels:
      app: test1
  template:
    metadata:
      labels:
        app: test1
    spec:
      containers:
      - env:
        - name: S6_READ_ONLY_ROOT
          value: "1"
        image: public.ecr.aws/j1r0q0g6/notebooks/notebook-servers/jupyter-scipy:v1.3.0-rc.0
        imagePullPolicy: IfNotPresent
        name: jupyter-scipy
        volumeMounts:
        - mountPath: /var/run/s6
          name: s6
      volumes:
      - name: s6
        emptyDir: {}
```

**Comment by juliusvonkohout** [comment]
```
kind: Deployment
apiVersion: apps/v1
metadata:
  annotations:
    deployment.kubernetes.io/revision: '1'
  name: test1
  namespace: prod-1
spec:
  replicas: 1
  selector:
    matchLabels:
      app: test1
  template:
    metadata:
      labels:
        app: test1
        sidecar.istio.io/inject: 'false' # it also runs without but in this way i could easily create a service and route
    spec:
      serviceAccountName: default-editor
      securityContext: {}
      containers:
        - name: jupyter-scipy
          image: >-
            public.ecr.aws/j1r0q0g6/notebooks/notebook-servers/jupyter-scipy:v1.3.0-rc.0
          env:
            - name: S6_READ_ONLY_ROOT
              value: '1'
          resources: {}
          volumeMounts:
            - name: s6
              mountPath: /var/run/s6
      serviceAccount: default-editor
      volumes:
        - name: s6
          emptyDir: {}
```

````
(base) jovyan@test1-6fccf795ff-lhvpx:~$ id
uid=1000(jovyan) gid=100(users) groups=100(users),1001230000
```

So it is overwriting the default GID 0 with 100. I just have to test the file permissions if i mount some PVC at /home/jovyan.

**Comment by DavidSpek** [comment]
@juliusvonkohout Awesome, thanks for testing that. If you have the time and energy, could you also check if it breaks if you remove the `S6_READ_ONLY_ROOT=1` environment variable? Would like to be sure what exactly is needed so I can incorporate the needed changes in the controller and possibly the Dockerfiles.

**Comment by juliusvonkohout** [comment]
> @juliusvonkohout Awesome, thanks for testing that. If you have the time and energy, could you also check if it breaks if you remove the `S6_READ_ONLY_ROOT=1` environment variable? Would like to be sure what exactly is needed so I can incorporate the needed changes in the controller and possibly the Dockerfiles.

I updated my deployment instructions above. S6_READ_ONLY_ROOT=1 does not seem to be needed.

```
(base) jovyan@test1-5d7df684c8-7bmrf:~$ printenv | grep s6
OLDPWD=/run/s6/services/jupyterlab
(base) jovyan@test1-5d7df684c8-7bmrf:~$ printenv | grep S6
(base) jovyan@test1-5d7df684c8-7bmrf:~$ 
```

**Comment by DavidSpek** [comment]
Great, then the change necessary to the notebook-controller is very minimal. I'm not sure how the SCC / PSP can best be handled, as this would probably need to be included in the deployment manifests somewhere.

**Comment by juliusvonkohout** [comment]
> Great, then the change necessary to the notebook-controller is very minimal. I'm not sure how the SCC / PSP can best be handled, as this would probably need to be included in the deployment manifests somewhere.

Yes, this is completely out of scope for this PR.

**Comment by DavidSpek** [comment]
@juliusvonkohout Are you in the Kubeflow Slack? I'd like to discuss and coordinate some things with you regarding this and similar problems when using OpenShift.

**Comment by juliusvonkohout** [comment]
> @juliusvonkohout Are you in the Kubeflow Slack? I'd like to discuss and coordinate some things with you regarding this and similar problems when using OpenShift.

No, how can i get access?

**Comment by srikantt** [comment]
> 

how do I update the statefulset if it is getting generated by the notebook-controller using unknown parameters/arguments?

**Comment by tim5chneider** [comment]
I concur with @srikantt question! 
Has anybody a hint to fix this issue with a workaround?
Or is it possible that the manual pod deployment with this yml (https://github.com/kubeflow/kubeflow/issues/5808#issuecomment-851626585) is accessible from Kubeflow-GUI?

**Comment by juliusvonkohout** [comment]
> I concur with @srikantt question! Has anybody a hint to fix this issue with a workaround? Or is it possible that the manual pod deployment with this yml ([#5808 (comment)](https://github.com/kubeflow/kubeflow/issues/5808#issuecomment-851626585)) is accessible from Kubeflow-GUI?

Yes, build your own Jupyterlab in a proper way without s6.

**Comment by namlh16** [comment]
> Hi, I'am also experiencing the exact same issue while trying to deploy the `jupyter-scipy:v1.3.0-rc.1` Notebook Server via the KubeFlow 1.3 WebUI into the default `kubeflow-user-example-com` namespace on an OpenShift 4.7 cluster:
> 
> ```
> [s6-init] making user provided files available at /var/run/s6/etc...exited 0.
> [s6-init] ensuring user provided files have correct perms...s6-chown: fatal: unable to chown /var/run/s6/etc/cont-init.d/01-copy-tmp-home: Operation not permitted
> s6-chmod: fatal: unable to change mode of /var/run/s6/etc/cont-init.d/01-copy-tmp-home: Operation not permitted
> s6-chown: fatal: unable to chown /var/run/s6/etc/services.d/jupyterlab/run: Operation not permitted
> s6-chmod: fatal: unable to change mode of /var/run/s6/etc/services.d/jupyterlab/run: Operation not permitted
> exited 0.
> [fix-attrs.d] applying ownership & permissions fixes...
> [fix-attrs.d] done.
> [cont-init.d] executing container initialization scripts...
> [cont-init.d] 01-copy-tmp-home: executing...
> cp: cannot access '/tmp_home/jovyan/.jupyter': Permission denied
> [cont-init.d] 01-copy-tmp-home: exited 1.
> [cont-init.d] done.
> [services.d] starting services
> s6-supervise (child): fatal: unable to exec run: Permission denied
> s6-supervise jupyterlab: warning: unable to spawn ./run - waiting 10 seconds
> [services.d] done.
> s6-supervise (child): fatal: unable to exec run: Permission denied
> s6-supervise jupyterlab: warning: unable to spawn ./run - waiting 10 seconds
> ```
> 
> After a while the Notebook status changes to `Running` but trying to `CONNECT` results in a 503 `upstream connect error or disconnect/reset before headers. reset reason: connection failure`.

im face to same issue with kubeflow & openshift. Are you solve this problem ?, please help me
@andreas-schliebitz

**Comment by themowski** [comment]
Our team encountered the error discussed here, but in a slightly different context (I think). We were able to resolve the issue, read on for more information.

# **tl,dr;**

Our team has to use a custom `securityContext` in order to access an NFS share. This prevents us from running as the default jovyan user (our UID/GID are different).

We ran `chmod 775 s6/services.d/jupyterlab/run` before building our custom Jupyter image. This corresponds to the file https://github.com/kubeflow/kubeflow/blob/master/components/example-notebook-servers/jupyter/s6/services.d/jupyterlab/run here in the kubeflow/kubeflow repo.

We also added `chmod -R 775 /tmp_home` to the relevant place in the `jupyter` Dockerfile while building our custom image, but it is not clear whether that had any impact on fixing the issue.

# Problem Description / Environment

We created a custom notebook image (called `custom-jupyter:1` in this discussion) based on the `base`, `jupyter`, and `jupyter-tensorflow` Dockerfiles. Our custom Dockerfile combines the key parts of these into one, and reorganizes some of the steps slightly, but overall it is essentially the same logic condensed into a single file. We were able to spin up an instance of this custom image and access JupyterLab without problems.

We are using Kubeflow 1.4.1 on Kubernetes 1.21.9. The Kubernetes cluster was provisioned through Rancher v2.6.3-patch1. We are using Docker 20.10.12 as our container engine under the hood.

We need to mount an NFS share, and in order to do that, we need to set a `securityContext` that changes the UID & GID. We currently do this by injecting the `securityContext` block and the relevant NFS volume mounts via `kubectl patch`. For reference, assume we spun up a notebook called `test`, which creates pod `test-0`; here is a heavily-redacted example of what `kubectl get -n myuser -o yaml pod test-0` gives us, showcasing the injected data in the `securityContext`, `volumeMounts`, and `volumes` areas:

```yaml
(...)
spec:
  containers:
  - (...)
    image: custom-jupyter:1
    name: test
    (...)
    securityContext:
      runAsGroup: 1234
      runAsUser: 1234
    (...)
    volumeMounts:
    - mountPath: /home/jovyan
      name: workspace-test
    - mountPath: /mnt/nfs_share
      name: my-nfs
    (...)
    workingDir: /home/jovyan
  (...)
  volumes:
  (...)
  - name: my-nfs
    nfs:
      path: /data
      server: my.nfs.server.path
(...)
```

After patching our `custom-jupyter:1` notebook image, we are unable to connect to JupyterLab. As indicated throughout this thread, clicking the "CONNECT" button in the Kubeflow UI takes us to a page that just says:

```text
upstream connect error or disconnect/reset before headers. reset reason: connection failure
```

Our `kubectl logs` output is also similar to the output shown above:

```text
$ kubectl logs -n myuser test-0 test
[s6-init] making user provided files available at /var/run/s6/etc...exited 0.
s6-chown: fatal: unable to chown /var/run/s6/etc/cont-init.d/01-copy-tmp-home: Operation not permitted
s6-chmod: fatal: unable to change mode of /var/run/s6/etc/cont-init.d/01-copy-tmp-home: Operation not permitted
s6-chown: fatal: unable to chown /var/run/s6/etc/services.d/jupyterlab/run: Operation not permitted
s6-chmod: fatal: unable to change mode of /var/run/s6/etc/services.d/jupyterlab/run: Operation not permitted
[s6-init] ensuring user provided files have correct perms...exited 0.
[fix-attrs.d] applying ownership & permissions fixes...
[fix-attrs.d] done.
[cont-init.d] executing container initialization scripts...
[cont-init.d] 01-copy-tmp-home: executing...
cp: cannot access '/tmp_home/jovyan/.jupyter': Permission denied
[cont-init.d] 01-copy-tmp-home: exited 1.
[cont-init.d] done.
[services.d] starting services
s6-supervise (child): fatal: unable to exec run: Permission denied
s6-supervise jupyterlab: warning: unable to spawn ./run - waiting 10 seconds
[services.d] done.
s6-supervise (child): fatal: unable to exec run: Permission denied
s6-supervise jupyterlab: warning: unable to spawn ./run - waiting 10 seconds

((This error then repeats every 10 seconds, forever))
```

# Problem Solution

Mounting an `emptyDir` to `/var/run/s6` as suggested above did NOT fix the problem.

I zeroed in on those initial error messages from s6 -- specifically, I noticed that the permissions change of the `run` file for JupyterLab failed. I went back to our source for the Docker image and saw that `s6/services.d/jupyterlab/run` was not executable.  This file was copied as-is from the kubeflow/kubeflow repo's `jupyter` image example: https://github.com/kubeflow/kubeflow/blob/2d347e97d37b290d5764e84fc26f4d9870ba06ce/components/example-notebook-servers/jupyter/s6/services.d/jupyterlab/run (this URL is pinned to the latest master commit; the file is the same as the one we're using).

My assumption is that JupyterLab was failing to start because s6 failed to make the `run` script executable, and it wasn't already executable to begin with.

To fix this, I ran (in my local development environment)

```bash
cd custom-jupyter-docker/s6/services.d/jupyterlab/
chmod 775 run
```

That made the file executable.

I also noticed this error in the `kubectl logs`:

```text
[cont-init.d] 01-copy-tmp-home: executing...
cp: cannot access '/tmp_home/jovyan/.jupyter': Permission denied
[cont-init.d] 01-copy-tmp-home: exited 1.
```

After some investigation, it looks like after Jupyter gets installed, that particular directory has `600` permissions. **I don't know if this is even relevant to the issue at hand,** but I also updated the Dockerfile to set mode `775` on everything in `/tmp_home`. (I probably should/could do something like 664 instead, but this doesn't seem to have broken anything, so I probably will leave it as-is.)

```dockerfile
RUN mkdir -p /tmp_home \
 && cp -r ${HOME} /tmp_home \
 && chown -R ${NB_USER}:users /tmp_home \
 && chmod -R 775 /tmp_home
```

This change corresponds to lines 70-72 here: https://github.com/kubeflow/kubeflow/blob/2d347e97d37b290d5764e84fc26f4d9870ba06ce/components/example-notebook-servers/jupyter/Dockerfile

After this, I rebuilt the `custom-jupyter` Docker image. Deploying an instance of this new image along with the patches described above fixed the issue: I am able to access JupyterLab, and everything seems to work normally, despite the fact that I am running as a non-jovyan user with a different UID/GID.

I don't 100% understand why this works, but I think that since Istio injects the GID 1337 into my user as an additional GID (so I have both my custom 1234 GID shown in my sample YAML, and the Istio 1337 GID), the fact that everything is group-accessible allows me to have full access to jovyan's things, despite not being jovyan.

# Recommendation

My recommendation for now would be for the Kubeflow dev team to mark the `components/example-notebook-servers/jupyter/s6/services.d/jupyterlab/run` file as executable and then commit that to the repo and rebuild the sample images. Consider also making the change to `/tmp_home`, although as stated it is not clear that that does anything useful.

**Comment by Chenxs1122** [comment]
> securityContext:
>           runAsGroup: 100
>           runAsUser: 1000

the permissions of this mount directory will change

**Comment by jbottum** [comment]
@kubeflow/wg-notebooks-leads seems like the Notebooks WG might want to look/comment on this.

**Comment by Chenxs1122** [comment]
> @kubeflow/wg-notebooks-leads seems like the Notebooks WG might want to look/comment on this.

sorry!  the page is not found

**Comment by juliusvonkohout** [comment]
"the fact that everything is group-accessible allows me to have full access to jovyan's things, despite not being jovyan." this is how you are supposed to build all OCI images. Always use group 0 such that it can run as any user.

**Comment by kimkihoon0515** [comment]
@juliusvonkohout Hi, I have a question.
What I'm trying to do is using elyra on kubeflow.
I created a notebook through an image combining jupyterhub and ldap.
Users created through jupyterhub are created as local groups.
```
gid:501(local) groups:501(local) 
```
But the service account token which I made through poddefaults is in 1337 group just like this
```
drwxrwsrwt. 3 root 1337  100 Feb 24 07:50 .
drwxr-xr-x. 3 root root 4096 Feb 24 07:50 ..
drwxr-sr-x. 2 root 1337   60 Feb 24 07:50 ..2023_02_24_07_50_33.599224484
lrwxrwxrwx. 1 root 1337   31 Feb 24 07:50 ..data -> ..2023_02_24_07_50_33.599224484
lrwxrwxrwx. 1 root 1337   12 Feb 24 07:50 token -> ..data/token
```
so when i run `cat token` , it always occur permission denied error. Is there any solution for this problem??

**Comment by juliusvonkohout** [comment]
@kimkihoon0515 how is your question related to this issue? 

 Is there any solution for this problem??

Yes order consulting from me or other companies.

**Comment by prasannakumar-coupa** [comment]
Based on @juliusvonkohout solution, we can have PodDefault with volumeMounts and volumes https://github.com/kubeflow/kubeflow/blob/master/components/admission-webhook/README.md#how-this-works


```
apiVersion: kubeflow.org/v1alpha1
kind: PodDefault
metadata:
  name: add-s6-volume
  namespace: <user-namespace>
spec:
 selector:
  matchLabels:
    add-s6-volume: "true"
 desc: "add s6 volume"
  volumeMounts:
    - name: s6
      mountPath: /var/run/s6
  volumes:
    - name: s6
      emptyDir: {}
```

And it works.

**Comment by thesuperzapper** [comment]
Hey everyone, I have made a PR which will fix this issue in 1.9.0:

- https://github.com/kubeflow/kubeflow/pull/7622

Please take a look at that PR for test images if you want to try them sooner.
